### PR TITLE
Added KT.broadcast_left_join_distinct and KT.distinct_by_key.

### PR DESCRIPTION
**Pull request by cseed**

### CODE REVIEW

File: python/hail/keytable.py [hunk 1]
```diff
@@ -332,6 +332,44 @@ def join(self, right, how='inner'):
...
+        :return: Key table with keys the distinct keys of this key table
+        :rtype: :class:`.KeyTable`
+
+        """
+        
+        return KeyTable(self.hc, self._jkt.distinctByKey())
+
+    @handle_py4j
+    @typecheck_method(right=kt_type)
+    def broadcast_left_join_distinct(self, right):
```
Comment (tpoterba) [inline python/hail/keytable.py]: I think it's not a good idea to add methods with the same functionality but different implementation. I would prefer to add a flag onto join - left and inner are both easy (with and without distinct), and we can error out on unsupported modes.

File: src/main/scala/is/hail/keytable/KeyTable.scala [hunk 2]
```diff
@@ -484,6 +489,72 @@ case class KeyTable(hc: HailContext, rdd: RDD[Row],
...
+    val newSignature = TStruct((keySignature.fields ++ valueSignature.fields ++ right.valueSignature.fields)
+      .map(fd => (fd.name, fd.typ)): _*)
+    val localNKeys = nKeys
+    val size1 = valueSignature.size
+    val size2 = right.valueSignature.size
+    val totalSize = newSignature.size
+
+    assert(totalSize == localNKeys + size1 + size2)
+
+    val merger = (k: Row, r1: Row, r2: Row) => {
```
Comment (tpoterba) [inline src/main/scala/is/hail/keytable/KeyTable.scala]: The code copied from `join` makes me feel even more strongly that the two methods should be unified in `join`.

File: python/hail/keytable.py (hunk 1 above)
Comment (cseed) [reply python/hail/keytable.py]: This is a good idea.

I can't say I'm super jazzed by `...join(..., broadcast_right=True, right_make_distinct=True)` when we have a `distinct` operation on `KeyTable`s.  Maybe I won't handle distinct and we'll leave it for the query optimizer.

**Comment by cseed** [comment]
Will revise based on comments and resubmit.

File: python/hail/keytable.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/keytable.py]: Yeah, I think the join shouldn't have a distinct option. It should do a cartesian product like the others.

**Comment by tpoterba** [comment]
never revised and resubmitted :(

**Pull request closed without merging**